user/applications/tinker/src/board_config.cpp

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2024 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "board_config.h"
+
+#ifdef ENABLE_MUON_DETECTION
+
+namespace {
+    JSONValue getValue(const JSONValue& obj, const char* name) {
+        JSONObjectIterator it(obj);
+        while (it.next()) {
+            if (it.name() == name) {
+                return it.value();
+            }
+        }
+        return JSONValue();
+    }
+
+    bool has(const JSONValue& obj, const char* name) {
+        return getValue(obj, name).isValid();
+    }
+}
+
+namespace particle {
+
+BoardConfig::BoardConfig()
+        : replyWriter_(nullptr, 0),
+          isMuon_(false),
+          isMuonDetected_(false) {
+}
+
+BoardConfig::~BoardConfig() {
+}
+
+BoardConfig* BoardConfig::instance() {
+    static BoardConfig config;
+    return &config;
+}
+
+bool BoardConfig::process(JSONValue config) {
+    memset(replyBuffer_, 0x00, sizeof(replyBuffer_));
+    replyWriter_ = JSONBufferWriter((char*)replyBuffer_, sizeof(replyBuffer_));
+    if (has(config, "BASE_BOARD_DETECT")) {
+        detectBaseBoard();
+        return true;
+    } else if (has(config, "CONFIGURE_MODULE_BOARD")) {
+        auto value = getValue(config, "CONFIGURE_MODULE_BOARD");
+        configureBaseBoard(value);
+        return true;
+    }
+    return false;
+}
+
+char* BoardConfig::reply() {
+    return replyWriter_.buffer();
+}
+
+size_t BoardConfig::replySize() {
+    return replyWriter_.dataSize();
+}
+
+void BoardConfig::detectBaseBoard() {
+    detectI2cSlaves();
+    replyWriter_.beginObject();
+    if (isMuon_) {
+        replyWriter_.name("board").value("muon");
+    } else {
+        replyWriter_.name("board").value("none");
+    }
+    replyWriter_.endObject();
+}
+
+void BoardConfig::configureBaseBoard(JSONValue value) {
+    detectI2cSlaves(false);
+    int ret;
+    if (isMuon_ && value.toString() == "muon") {
+        ret = configure(true);
+    } else if (!isMuon_ && value.toString() == "none") {
+        ret = configure(false);
+    } else {
+        ret = SYSTEM_ERROR_INVALID_ARGUMENT;
+    }
```

Comment:
Having the `isMuon` and `!isMuon` check will prevent resetting a module back to default values and will make end-to-end validation harder.

Here's the scenario I'm thinking of:
1. Take a new M-SoM out of the box and plug it into a Muon. At that moment the M-SoM is configured for "none".
2. Run through setup and check that the module was properly configured for "muon".
3. Before doing another test of setup, reset the M-SoM to "none" using a script. However you'll get invalid argument, so it's not possible to repeat steps 1 and steps 2 as if your device was a new M-SoM fresh out of the box.

Reply:
Makes sense. This flag was used to avoid wrongly configuring those pins, but if the setup page can ensure that the correct JSON command is sent, then we're fine to remove the check.

Reply:
Let's remove this check, it's better not to restrict this call to set configuration to non-muon.